Editable annotation cards in itemPane

abaevbog opened the pull request:
- tags in the annotation cards will open a popup with tags section, same as in the reader. The functionality of creating and opening the popup is shared in new `Zotero.Annotations.insertAnnotationsTagsPopup`.
- options button in the top-right corner of the annotation card will open a menu with colors and "Delete" option
- keyboard navigation somewhat similar to one in the sidenav of the reader: arrowUp/Down will navigate
between annotations, tab will navigate through annotation comment and tags. Next tab will move focus out of the annotations list.
- annotation comment can be edited. On click, `simpleEditor.html` iframe will be inserted instead of the comment div. There, one has rich text editor capabilities, same as in the reader. `simpleEditor.html` iframe is used because `contenteditable` in the root chrome window does not work properly (this issue must have been the reason why the original `quickFormat` dialog had the `iframe` too) . Edits are saved on blur, or on Cmd/Ctrl+Enter. Escape cancels the edits. 
- refactoring of `simpleEditor` to allow for easier customization via attributes placed on the iframe. Possible attributes are: `actions` (to specify which actions should appear), `tight` (to indicate that the buttons should be smaller than default), and `toolbar-below-editor` (to indicate where toolbar should be placed). This way, simpleEditor can be customized here without affecting editBibliographyDialog. For convenience, moved simpleEditor stylesheet into new _simpleEditor.scss. I expect this customization will also help in the redesign of the `editBibliographyDialog`.
- do not set `aria-expanded` on non-expandable header of `collapsible-section`, since that implies the section can be expanded.

Notes: 
- marking this as a draft for now until https://github.com/zotero/zotero/issues/5247 is resolved. Currently, we strip all html tags from annotations. But https://github.com/zotero/zotero/pull/4607 should allow us to properly render them safely and perform any conversions necessary, as in the reader. Functions like `_commentFromHTML` and `_commentToHTML` of  `annotation-row` would likely be affected by it too, and for now are just minimal implementations to get some editing functionality working. 
- there is also a question about editable and delete-able annotations discussed in #5262. In reader, external annotations or annotations made by other users cannot be edited or deleted. However, external annotations are editable and delete-able on the `item.js` level. Annotations made by other users can not be edited but still can be deleted. I am not exactly sure about the intended behavior here. For now, external and other user's annotation cards cannot be interacted with, same as in the reader. Backspace on external annotation rows also will not delete them but backspace on group annotation rows will (as this is explicitly done in https://github.com/zotero/zotero/commit/df64a16b55c17264949f30cfc9edeb4c098f169a). Correct me if this is off.

Fixes: #5262
Fixes: #5229

abaevbog committed:
editable annotation cards in itemPane

- tags in the annotation cards will open a popup with
tags section, same as in the reader. The functionality
of creating and opening the popup is shared in
new Zotero.Annotations.insertAnnotationsTagsPopup.
- options button in the top-right corner of the annotation
card will open a menu with colors and "Delete" option
- keyboard navigation somewhat similar to one in the
sidenav of the reader: arrowUp/Down will navigate
between annotations, tab will navigate through annotation
comment and tags. Next tab will move focus out of the
annotations list.
- annotation comment can be edited. On click, simpleEditor.html
iframe will be inserted instead of the comment div. There,
one has rich text editor capabilities, same as in the reader.
simpleEditor.html iframe is used because contenteditable
in the root chrome window does not work properly. Edits are
saved on blur, or on Cmd/Ctrl+Enter. Escape cancels the edits.
- refactoring of simpleEditor to allow for easier customization
via attributes placed on the iframe. Possible attributes are:
actions (to specify which actions should appear),
tight (to indicate that the buttons should be smaller than default),
and toolbar-below-editor (to indicate where toolbar should be placed).
This way, simpleEditor can be customized here without affecting
editBibliographyDialog. For convenience, moved simpleEditor
stylesheet into new _simpleEditor.scss.

Minor additional edits:
- backspace on external annotations will not delete them.
Annotations can still technically be erased if the
parent item is erased
- do not set aria-expanded on non-expandable header
of collapsible-section

abaevbog committed:
fix no html tags appearing in comment

windingwind commented:
That sounds natural. Thanks for making this.

I have something about annotations in the item pane in mind, as we discussed in #5241, and I think the easiest way is to use the current `ItemDetails` for single annotation selection, where other custom sections can automatically work; for multiple annotation selection, it uses the `annotationsItemPane`.

The thing is, do we want to use the already-there tags box in the `ItemDetails` if only one annotation is selected? I suppose no, and we stick to the implementation here to edit the annotation tags, right?

abaevbog commented:
Gotcha, yeah that makes sense! Thank you for the explanation. 

I see that `itemDetails` has all the logic to handle custom sections. Do I understand correctly that the idea is to avoid duplicating that logic in `annotationItemPane`?

>  if any custom section for single annotation selection is added, the single annotation selection item pane becomes the way it works with other item types, like with several item pane sections and collapsible/pinnable.

What about custom sections for when multiple annotations are selected? Is that something we are not considering at the moment?

abaevbog commented:
> Yes, and besides that, the custom sections and the API by design don't accept multiple items

Oh, I see! That makes sense. It actually clarifies the main question for me.

I kept thinking about multiple items being selected because it is really the main value of the having annotation rows in itemTree (e.g. so you can find and do something to all annotations fitting some criteria across many items). But if custom sections for multiple annotations is not something we are after at the moment, using `itemDetails` for a single selection makes perfect sense to me, thank you for the explanation!

abaevbog committed:
smoother transition to the editor on click

abaevbog committed:
fix stylesheet conflicts

abaevbog committed:
better read-only mode

For when an annotation itself is editable but we don't
want to allow it at some specific instace.

abaevbog committed:
make annotation rows selectable

- keyboard navigation with shift+arrow support for range
selection
- click on a row will select it, Cmd/Ctrl+click will toggle
selection, Shift+click will select a range

abaevbog committed:
can drag selected annotation-rows rows into editor

For quickCopy

abaevbog committed:
trigger the menu via right-click as well

And apply selected action (color change, deletion)
to all selected annotations

abaevbog committed:
added create/add note option to context menu

abaevbog committed:
keep focus on the actual annotation row on click

abaevbog committed:
open context menu at mouse position on right click
